src/vs/workbench/contrib/terminal/electron-browser/terminalInstance.ts

```diff
@@ -739,14 +739,33 @@ export class TerminalInstance implements ITerminalInstance {
 		}
 		if (this._processManager) {
 			this._processManager.dispose(immediate);
+		} else {
+			// In cases where there is no associated process (for example executing an extension callback task)
+			// consumers still expect on onExit event to be fired. An example of this is terminating the extension callback
+			// task. There is no exit code at this point, so firing undefined is appropriate.
+			this._onExit.fire(undefined);
```

Comment:
It's a little weird that onExit is called when there is no exit code, why is this needed exactly? What do consumers do with it?

Reply:
The task system uses relies on this to know when a task has completed. It uses this to drive task state changes.

So, if this doesn't fire, then the task system believes that the task never completed.

Reply:
The task system uses relies on this to know when a task has completed. It uses this to drive task state changes.

So, if this doesn't fire, then the task system believes that the task never completed.